File: merge_sort_c++/mergesort/merge_sort.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <random>   // Incluye la biblioteca random
#include <chrono>   // Incluye la biblioteca chrono

void merge(std::vector<int>& arr, int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    std::vector<int> left_half(n1);
    std::vector<int> right_half(n2);

    for (int i = 0; i < n1; ++i)
        left_half[i] = arr[left + i];
    for (int j = 0; j < n2; ++j)
        right_half[j] = arr[mid + 1 + j];

    int i = 0;
    int j = 0;
    int k = left;

    while (i < n1 && j < n2) {
        if (left_half[i] <= right_half[j]) {
            arr[k] = left_half[i];
            ++i;
        }
        else {
            arr[k] = right_half[j];
            ++j;
        }
        ++k;
    }

    while (i < n1) {
        arr[k] = left_half[i];
        ++i;
        ++k;
    }

    while (j < n2) {
        arr[k] = right_half[j];
        ++j;
        ++k;
    }
}

void merge_sort(std::vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;

        merge_sort(arr, left, mid);
        merge_sort(arr, mid + 1, right);

        merge(arr, left, mid, right);
    }
}

int main() {
    for (int i = 0; i < 5; i++)
    {
        //repetir el codigo 5 veces

        std::srand(static_cast<unsigned int>(std::time(nullptr)));

        int n = 50000; // Tamaño de la lista
        std::vector<int> unsorted_list;


        // Crea un generador de números aleatorios usando una semilla adecuada
        std::random_device rd;
        std::mt19937 gen(rd());

        // Crea una distribución uniforme de números en un rango
        std::uniform_int_distribution<int> dis(1, 10000); // Cambia el rango según tus necesidades

        // Genera números aleatorios y agrega a la lista
        for (int i = 0; i < n; ++i) {
            unsorted_list.push_back(dis(gen));
        }


        // Imprime la lista generada
        /*
        std::cout << "Lista aleatoria: ";
        for (int num : unsorted_list) {
            std::cout << num << " ";
        }*/

        // Captura el tiempo de inicio
        auto start_time = std::chrono::high_resolution_clock::now();

        merge_sort(unsorted_list, 0, unsorted_list.size() - 1);

        // Captura el tiempo de finalización
        auto end_time = std::chrono::high_resolution_clock::now();

        // Calcula la duración de la ejecución
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time);

        // Imprime la lista ordenada
        /*
        std::cout << "\nLista ordenada:";
        for (int num : unsorted_list) {
            std::cout << " " << num;
        }*/
        std::cout << "\nTiempo de ejecución: " << duration.count() << " microsegundos" << std::endl;
    }
    std::cout << std::endl;

    return 0;
}

File: practica 1/merge_sort.py
import random
import time

def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]

    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)

    return merge(left_half, right_half)

def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    
    return result

def generate_random_list(n, min_val, max_val):
    return [random.randint(min_val, max_val) for _ in range(n)]

# Ejemplo de uso:

tiempo_inicial=time.time()
unsorted_list = generate_random_list(50000, 0, 10000)
print("Lista original:", unsorted_list)
sorted_list = merge_sort(unsorted_list)
print("Lista ordenada:",sorted_list)
tiempo_final = time.time()
execution_time=tiempo_final-tiempo_inicial
print("Tiempo de ejecución:", execution_time, "segundos")
